### DISCUSSION TITLE
"use client" is a bit confusing

### DISCUSSION THREAD

**Discussion by kecoliva** [vercel/next.js#60646] [Help]
### Summary

If I want to call `useEffect` within my page (e.g. I want to show a toast message on page render), I have two options right?

1. Make the page a client component by adding `"use client"` on top of the page and then calling the `useEffect` hook directly within the page.
2. Create a client component dedicated to calling the `useEffect` hook to render the toast message, with this I won't be necessarily making the whole page a client component in result better performance?

Is my understanding here right? That option 2 is better but at the same time it would be weird to create a component just to call one `useEffect` hook and that wouldn't be called anywhere.

Option1
![image](https://github.com/vercel/next.js/assets/46006869/f6072815-20ba-4be3-84ab-6f65cccfdeec)

Option2
![image](https://github.com/vercel/next.js/assets/46006869/71543595-ec98-4567-8803-7b4cf3eef287)
![image](https://github.com/vercel/next.js/assets/46006869/cfb32055-1562-4d51-b5ae-adddc305ae19)

**Comment by icyJoseph** [comment]
Not in this case. I think your option 2 would crash.

The problem here is that you'd call the hook during the server component render step, and that'd encounter a client module. I think one can almost think of the Server Component render step as a way to create a stringified VDOM, where everything that's a client module, is effectively a hole to be filled in by the Client Component tree. 

So calling a hook while rendering a server component (before the return statement), effectively, requires that server component to be switched to a client component. What you need to do is create `ToastRenderer`:

```jsx
"use client";
import { useEffect } from "react";

export const ToastRenderer = () => {
  useEffect(() => {
    // toast render logic
  });
  return null;
};
```

And use it like:

```jsx
// server component page
import { ToastRenderer } from "./useSomeEffect";

export default function Home() {
  return (
    <>
      <ToastRenderer />

      <main>
        <h1>My UI</h1>
      </main>
    </>
  );
}
```
